### PR TITLE
Merge upstream prometheus/prometheus at `85e3c43`

### PR DESCRIPTION
**Pull request by gotjosh**
# Depedencies
- https://github.com/prometheus/prometheus/pull/14029
- https://github.com/prometheus/prometheus/pull/14035
- https://github.com/prometheus/prometheus/pull/14024
- https://github.com/prometheus/prometheus/pull/14027
- https://github.com/prometheus/prometheus/pull/14031
- https://github.com/prometheus/prometheus/pull/14037
# Relevant
- https://github.com/prometheus/prometheus/pull/13980 (The PR I'm doing this update for)
# Not Relevant
- https://github.com/prometheus/prometheus/pull/14012
- https://github.com/prometheus/prometheus/pull/14014

### CODE REVIEW

**Comment by gotjosh** [comment]
I had a set of conflicts:

```
Unmerged paths:
  (use "git add/rm <file>..." as appropriate to mark resolution)
        deleted by us:   .github/workflows/buf-lint.yml
        deleted by us:   .github/workflows/buf.yml
        deleted by us:   .github/workflows/ci.yml
        deleted by us:   .github/workflows/codeql-analysis.yml
        both modified:   .github/workflows/container_description.yml
        deleted by us:   .github/workflows/fuzzing.yml
        deleted by us:   .github/workflows/repo_sync.yml
        deleted by us:   .github/workflows/scorecards.yml
        both modified:   go.mod
        both modified:   rules/manager_test.go
```

And the diff for the `go.mod` and `manager_test.go` is:

```git
diff --cc go.mod
index 21eab8786,14bdd93a0..000000000
--- a/go.mod
+++ b/go.mod
@@@ -93,14 -93,9 +93,14 @@@ require 
  	k8s.io/klog/v2 v2.120.1
  )
  
 +require (
 +	github.com/DmitriyVTitov/size v1.5.0
 +	github.com/dgraph-io/ristretto v0.1.1
 +)
 +
  require (
- 	cloud.google.com/go/auth v0.2.0 // indirect
- 	cloud.google.com/go/auth/oauth2adapt v0.2.0 // indirect
+ 	cloud.google.com/go/auth v0.3.0 // indirect
+ 	cloud.google.com/go/auth/oauth2adapt v0.2.2 // indirect
  	cloud.google.com/go/compute/metadata v0.3.0 // indirect
  	github.com/Azure/azure-sdk-for-go/sdk/internal v1.5.2 // indirect
  	github.com/AzureAD/microsoft-authentication-library-for-go v1.2.2 // indirect
diff --cc rules/manager_test.go
index 4b884ffed,07159145f..000000000
--- a/rules/manager_test.go
+++ b/rules/manager_test.go
@@@ -435,94 -424,92 +429,91 @@@ func TestForStateRestore(t *testing.T) 
  			num:             2,
  		},
  		{
- 			// No active alerts.
+ 			name:            "no active alerts",
  			restoreDuration: 50 * time.Minute,
- 			alerts:          []*Alert{},
+ 			expectedAlerts:  []*Alert{},
  		},
- 	}
- 
- 	testFunc := func(tst testInput, evalDelay time.Duration) {
- 		newRule := NewAlertingRule(
- 			"HTTPRequestRateLow",
- 			expr,
- 			alertForDuration,
- 			0,
- 			labels.FromStrings("severity", "critical"),
- 			labels.EmptyLabels(), labels.EmptyLabels(), "", false, nil,
- 		)
- 		newGroup := NewGroup(GroupOptions{
- 			Name:            "default",
- 			Interval:        time.Second,
- 			Rules:           []Rule{newRule},
- 			ShouldRestore:   true,
- 			Opts:            opts,
- 			EvaluationDelay: &evalDelay,
- 		})
- 
- 		newGroups := make(map[string]*Group)
- 		newGroups["default;"] = newGroup
- 
- 		restoreTime := baseTime.Add(tst.restoreDuration).Add(evalDelay)
- 		// First eval before restoration.
- 		newGroup.Eval(context.TODO(), restoreTime)
- 		// Restore happens here.
- 		newGroup.RestoreForState(restoreTime)
- 
- 		got := newRule.ActiveAlerts()
- 		for _, aa := range got {
- 			require.Zero(t, aa.Labels.Get(model.MetricNameLabel), "%s label set on active alert: %s", model.MetricNameLabel, aa.Labels)
- 		}
- 		sort.Slice(got, func(i, j int) bool {
- 			return labels.Compare(got[i].Labels, got[j].Labels) < 0
- 		})
- 
- 		// Checking if we have restored it correctly.
- 		switch {
- 		case tst.noRestore:
- 			require.Len(t, got, tst.num)
- 			for _, e := range got {
- 				require.Equal(t, e.ActiveAt, restoreTime)
- 			}
- 		case tst.gracePeriod:
- 			require.Len(t, got, tst.num)
- 			for _, e := range got {
- 				require.Equal(t, opts.ForGracePeriod, e.ActiveAt.Add(alertForDuration).Sub(restoreTime))
- 			}
- 		default:
- 			exp := tst.alerts
- 			require.Equal(t, len(exp), len(got))
- 			sortAlerts(exp)
- 			sortAlerts(got)
- 			for i, e := range exp {
- 				require.Equal(t, e.Labels, got[i].Labels)
- 
- 				// Difference in time should be within 1e6 ns, i.e. 1ms
- 				// (due to conversion between ns & ms, float64 & int64).
- 				activeAtDiff := evalDelay.Seconds() + float64(e.ActiveAt.Unix()+int64(tst.downDuration/time.Second)-got[i].ActiveAt.Unix())
- 				require.Equal(t, 0.0, math.Abs(activeAtDiff), "'for' state restored time is wrong")
- 			}
- 		}
- 	}
- 
- 	for _, evalDelay := range []time.Duration{0, time.Minute} {
- 		for _, tst := range tests {
- 			testFunc(tst, evalDelay)
- 		}
- 	}
- 
- 	// Testing the grace period.
- 	for _, duration := range []time.Duration{10 * time.Minute, 15 * time.Minute, 20 * time.Minute} {
- 		evalTime := baseTime.Add(duration)
- 		group.Eval(context.TODO(), evalTime)
- 	}
- 
- 	for _, evalDelay := range []time.Duration{0, time.Minute} {
- 		testFunc(testInput{
+ 		{
+ 			name:            "test the grace period",
  			restoreDuration: 25 * time.Minute,
- 			alerts:          []*Alert{},
+ 			expectedAlerts:  []*Alert{},
  			gracePeriod:     true,
- 			num:             2,
- 		}, evalDelay)
+ 			before: func() {
+ 				for _, duration := range []time.Duration{10 * time.Minute, 15 * time.Minute, 20 * time.Minute} {
+ 					evalTime := baseTime.Add(duration)
+ 					group.Eval(context.TODO(), evalTime)
+ 				}
+ 			},
+ 			num: 2,
+ 		},
+ 	}
 -
+ 	for _, tt := range tests {
+ 		t.Run(tt.name, func(t *testing.T) {
+ 			if tt.before != nil {
+ 				tt.before()
+ 			}
+ 
+ 			newRule := NewAlertingRule(
+ 				"HTTPRequestRateLow",
+ 				expr,
+ 				alertForDuration,
+ 				0,
+ 				labels.FromStrings("severity", "critical"),
+ 				labels.EmptyLabels(), labels.EmptyLabels(), "", false, nil,
+ 			)
+ 			newGroup := NewGroup(GroupOptions{
+ 				Name:          "default",
+ 				Interval:      time.Second,
+ 				Rules:         []Rule{newRule},
+ 				ShouldRestore: true,
+ 				Opts:          opts,
+ 			})
+ 
+ 			newGroups := make(map[string]*Group)
+ 			newGroups["default;"] = newGroup
+ 
+ 			restoreTime := baseTime.Add(tt.restoreDuration)
+ 			// First eval before restoration.
+ 			newGroup.Eval(context.TODO(), restoreTime)
+ 			// Restore happens here.
+ 			newGroup.RestoreForState(restoreTime)
+ 
+ 			got := newRule.ActiveAlerts()
+ 			for _, aa := range got {
+ 				require.Zero(t, aa.Labels.Get(model.MetricNameLabel), "%s label set on active alert: %s", model.MetricNameLabel, aa.Labels)
+ 			}
+ 			sort.Slice(got, func(i, j int) bool {
+ 				return labels.Compare(got[i].Labels, got[j].Labels) < 0
+ 			})
+ 
+ 			// Checking if we have restored it correctly.
+ 			switch {
+ 			case tt.noRestore:
+ 				require.Len(t, got, tt.num)
+ 				for _, e := range got {
+ 					require.Equal(t, e.ActiveAt, restoreTime)
+ 				}
+ 			case tt.gracePeriod:
+ 
+ 				require.Len(t, got, tt.num)
+ 				for _, e := range got {
+ 					require.Equal(t, opts.ForGracePeriod, e.ActiveAt.Add(alertForDuration).Sub(restoreTime))
+ 				}
+ 			default:
+ 				exp := tt.expectedAlerts
+ 				require.Equal(t, len(exp), len(got))
+ 				sortAlerts(exp)
+ 				sortAlerts(got)
+ 				for i, e := range exp {
+ 					require.Equal(t, e.Labels, got[i].Labels)
+ 
+ 					// Difference in time should be within 1e6 ns, i.e. 1ms
+ 					// (due to conversion between ns & ms, float64 & int64).
+ 					activeAtDiff := float64(e.ActiveAt.Unix() + int64(tt.downDuration/time.Second) - got[i].ActiveAt.Unix())
+ 					require.Equal(t, 0.0, math.Abs(activeAtDiff), "'for' state restored time is wrong")
+ 				}
+ 			}
+ 		})
  	}
  }

```

File: .github/workflows/container_description.yml [hunk 1]
```diff
@@ -17,7 +17,7 @@ jobs:
     if: github.repository_owner == 'prometheus' || github.repository_owner == 'prometheus-community' # Don't run this workflow on forks.
     steps:
       - name: git checkout
-        uses: actions/checkout@v4
+        uses: actions/checkout@0ad4b8fadaa221de15dcec353f45205ec38ea70b # v4.1.4
```
Comment (gotjosh) [inline .github/workflows/container_description.yml:20]: This was the other conflict which is just updating the checkout action to a more recent version.

**Review by pstibrany — APPROVED** [review]
As always, please merge with a merge commit. Thank you!

**Review by pracucci — APPROVED** [review]
LGTM. As peter said, always merge sync PRs with merge commit 👇 

Screenshot 2024-05-02 at 15 04 13

**Pull request merged**